### PR TITLE
新增查看配置权限

### PR DESCRIPTION
**Pull request by SiouWang**
1.新增查看权限，新增后系统对配置的操作权限分为："查看配置", "修改配置", "发布配置";经过测试的"查看权限"有：
	1.1 按照不同环境授权用户是否可以查看配置
	1.2 按照不同集群授权用户是否可以查看配置
2.将发布按钮的显示权限由(namespace.hasReleasePermission || namespace.hasModifyPermission)改为(namespace.hasReleasePermission)，即拥有发布权限的时候才显示
3.将发布历史按钮新增了显示权限，修改为namespace.hasReleasePermission，即拥有发布权限的时候才允许查看发布历史

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,4 @@ target
 
 # git
 *.orig
+.idea*
```
Comment (nobodyiam) [inline .gitignore]: is it necessary?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/PermissionType.java [hunk 2]
```diff
@@ -1,27 +1,32 @@
...
 
+/**
+ * 权限类型
+ */
 public interface PermissionType {
 
+
   /**
-   * APP level permission
+   * 应用级别的权限
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/PermissionType.java]: I think it's ok to leave the comment as English, as it's our goal to make apollo readable to world wide developers

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/PermissionType.java [hunk 3]
```diff
@@ -1,27 +1,32 @@
...
 
   /**
    * 分配用户权限的权限
    */
   String ASSIGN_ROLE = "AssignRole";
 
+
   /**
-   * namespace level permission
+   * namespace级别的权限
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/PermissionType.java]: I think it's ok to leave the comment as English, as it's our goal to make apollo readable to world wide developers

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/PermissionType.java [hunk 4]
```diff
@@ -1,27 +1,32 @@
...
    */
-
   String MODIFY_NAMESPACE = "ModifyNamespace";
-
   String RELEASE_NAMESPACE = "ReleaseNamespace";
 
+  /**
+   * 查看配置的权限
+   */
+  String BROWSE_CONFIG = "BrowseConfig";
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/PermissionType.java]: I think it's better to name it as `VIEW_NAMESPACE` using the same format as `MODIFY_NAMESPACE` and `RELEASE_NAMESPACE `

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/RoleType.java [hunk 5]
```diff
@@ -2,14 +2,28 @@
...
 
+  /**
+   * 发布配置角色
+   */
   public static final String RELEASE_NAMESPACE = "ReleaseNamespace";
 
+  /**
+   * 配置查看角色
+   */
+  public static final String VIEW = "View";
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/constant/RoleType.java]: VIEW_NAMESPACE?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRoleInitializationService.java [hunk 6]
```diff
@@ -34,38 +35,58 @@
...
     initNamespaceEnvRoles(appId, ConfigConsts.NAMESPACE_APPLICATION, operator);
 
     //assign modify、release namespace role to user
-    rolePermissionService.assignRoleToUsers(
-        RoleUtils.buildNamespaceRoleName(appId, ConfigConsts.NAMESPACE_APPLICATION, RoleType.MODIFY_NAMESPACE),
-        Sets.newHashSet(operator), operator);
-    rolePermissionService.assignRoleToUsers(
-        RoleUtils.buildNamespaceRoleName(appId, ConfigConsts.NAMESPACE_APPLICATION, RoleType.RELEASE_NAMESPACE),
-        Sets.newHashSet(operator), operator);
+    rolePermissionService.assignRoleToUsers(RoleUtils.buildNamespaceRoleName(appId, ConfigConsts.NAMESPACE_APPLICATION, RoleType.MODIFY_NAMESPACE), Sets.newHashSet(operator), operator);
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRoleInitializationService.java]: looks like you haven't imported the apollo code style - https://github.com/ctripcorp/apollo/tree/master/apollo-buildtools/style

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultUserService.java [hunk 7]
```diff
@@ -34,6 +34,16 @@ public UserInfo findByUserId(String userId) {
...
   }
 
+  /**
+   * 查询所有用户列表
+   *
+   * @return
+   */
+  @Override
+  public List<UserInfo> selectUserList() {
+    return null;
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultUserService.java]: Arrays.asList(assembleDefaultUser())?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/ldap/LdapUserService.java [hunk 8]
```diff
@@ -90,4 +90,14 @@ public UserInfo findByUserId(String userId) {
...
   }
 
+  /**
+   * 查询所有用户列表
+   *
+   * @return
+   */
+  @Override
+  public List<UserInfo> selectUserList() {
+    return null;
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/ldap/LdapUserService.java]: Is this a bug?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/springsecurity/SpringSecurityUserService.java [hunk 9]
```diff
@@ -102,5 +102,21 @@ public UserInfo findByUserId(String userId) {
...
 
+  /**
+   * 查询所有用户列表
+   *
+   * @return
+   */
+  @Override
+  public List<UserInfo> selectUserList() {
+
+    List<UserPO> userList = userRepository.findAll();
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/springsecurity/SpringSecurityUserService.java]: why returning all users? this list could be very large, e.g. thousands of users

File: apollo-portal/src/main/resources/apollo-env.properties [hunk 10]
```diff
@@ -3,4 +3,4 @@ dev.meta=${dev_meta}
 fat.meta=${fat_meta}
 uat.meta=${uat_meta}
 lpt.meta=${lpt_meta}
-pro.meta=${pro_meta}
+pro.meta=http://localhost:8080
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/apollo-env.properties]: this should be reverted

File: apollo-portal/src/main/resources/static/scripts/directive/directive.js [hunk 11]
```diff
@@ -1,131 +1,129 @@
 /** navbar */
-directive_module.directive('apollonav',
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/scripts/directive/directive.js]: code style change?

File: apollo-portal/src/main/resources/static/views/common/nav.html [hunk 12]
```diff
@@ -25,10 +25,9 @@
                         <span class="glyphicon glyphicon-question-sign"></span> 帮助
                     </a>
                 </li>
-                <li class="dropdown" ng-if="hasRootPermission">
+                <li class="dropdown" >
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/common/nav.html]: why?

File: apollo-portal/src/main/resources/static/views/component/namespace-panel-master-tab.html [hunk 13]
```diff
@@ -40,22 +40,28 @@
 
             <div class="col-md-6 col-sm-6 text-right header-buttons">
 
+                <!-- 拥有发布权限的时候才显示发布按钮 -->
                 <button type="button" class="btn btn-success btn-sm"
                         data-tooltip="tooltip" data-placement="bottom" title="发布配置"
-                        ng-show="(namespace.hasReleasePermission || namespace.hasModifyPermission)"
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/namespace-panel-master-tab.html]: why changing this logic?

**Comment by nobodyiam** [comment]
非常感谢，有一些comment已经贴到相应代码段上了，另外，关于查看权限，目前的实现看上去是基于应用级别的？建议保持和发布/编辑逻辑一样吧，应用粒度对大部分情况或许可以，不过肯定有情况是需要到namespace级别的

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/repository/UserRepository.java [hunk 14]
```diff
@@ -2,6 +2,7 @@
 
 import com.ctrip.framework.apollo.portal.entity.po.UserPO;
 
+import org.springframework.data.jpa.repository.Query;
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/repository/UserRepository.java]: Unused import?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRoleInitializationService.java [hunk 15]
```diff
@@ -34,38 +35,58 @@
...
   @Transactional
   public void initAppRoles(App app) {
-    String appId = app.getAppId();
 
+    String appId = app.getAppId();
+    // 创建 App 拥有者的角色名
     String appMasterRoleName = RoleUtils.buildAppMasterRoleName(appId);
 
-    //has created before
+    // 校验角色是否已经存在。若是，直接返回
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRoleInitializationService.java]: Keep English comment

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRoleInitializationService.java [hunk 16]
```diff
@@ -23,6 +23,7 @@
 import java.util.*;
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRoleInitializationService.java]: Avoid using star imports

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/ctrip/CtripUserService.java [hunk 17]
```diff
@@ -96,6 +96,16 @@ public UserInfo findByUserId(String userId) {
...
   }
 
+  /**
+   * 查询所有用户列表
+   *
+   * @return
+   */
+  @Override
+  public List<UserInfo> selectUserList() {
+    return null;
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/ctrip/CtripUserService.java]: Will returning am empty list (`Collections.emptyList()`) better?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/UserInfoController.java [hunk 18]
```diff
@@ -75,5 +75,14 @@ public UserInfo getUserByUserId(@PathVariable String userId) {
     return userService.findByUserId(userId);
   }
 
+  /**
+   * 查询用户列表
+   * @return
+   */
+  @GetMapping("/users/list")
+  public List<UserInfo> getUserList() {
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/UserInfoController.java]: Do we really need a separate API? Does the API below meets the requirement?
```java

  @GetMapping("/users")
  public List<UserInfo> searchUsersByKeyword(@RequestParam(value = "keyword") String keyword,
                                             @RequestParam(value = "offset", defaultValue = "0") int offset,
                                             @RequestParam(value = "limit", defaultValue = "10") int limit) {
    return userService.searchUsers(keyword, offset, limit);
  }
```

such as `/users?keyword=*`?

If not, the API pattern doesn't respect to RESTful style, I suggest using `@GetMapping("/users")` and change the annotation of `searchUsersByKeyword ` to `  @GetMapping(value = "/user/logout", params = {"keyword"})`

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRolePermissionService.java [hunk 19]
```diff
@@ -198,7 +198,7 @@ public boolean userHasPermission(String userId, String permissionType, String ta
     }
 
     public boolean isSuperAdmin(String userId) {
-        return portalConfig.superAdmins().contains(userId);
+                return portalConfig.superAdmins().contains(userId);
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRolePermissionService.java]: Keep indentation to 4 spaces

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRolePermissionService.java [hunk 20]
```diff
@@ -230,7 +230,7 @@ public Permission createPermission(Permission permission) {
             Collection<String> permissionTypes = targetIdPermissionTypes.get(targetId);
             List<Permission> current =
                     permissionRepository.findByPermissionTypeInAndTargetId(permissionTypes, targetId);
-            Preconditions.checkState(CollectionUtils.isEmpty(current),
+             Preconditions.checkState(CollectionUtils.isEmpty(current),
```
Comment (kezhenxu94) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/defaultimpl/DefaultRolePermissionService.java]: Remove whitespace

File: apollo-portal/src/main/resources/apollo-env.properties (hunk 10 above)
Comment (kezhenxu94) [inline apollo-portal/src/main/resources/apollo-env.properties]: remove changes for development

**Pull request closed without merging**